lookup: add verify-node-gyp

bzoz opened the pull request:
Continued from https://github.com/nodejs/citgm/pull/252:

This adds testing if `node-gyp`  was called by `npm`. It uses `npm_config_node_gyp` environment variable  to inject our version of `node-gyp` script to detect if it was called with either `'build'` or `'rebuild'` parameter.

bzoz commented:
Rebased on master, PTAL

gdams commented:
@bzoz do you want to squash the commits and give it another rebase and i'll take another look

bzoz commented:
Rebased, PTAL

bzoz commented:
Fixed test.

MylesBorins commented on lib/citgm-node-gyp.js:
```diff
@@ -0,0 +1,23 @@
+
+var args = process.argv.slice(1);
+if (args.indexOf('build') !== -1 || args.indexOf('rebuild') !== -1) {
+  fs.writeFileSync(marker_filename, 'node-gyp called');
```
would prefer this not to be done with sync

bzoz replied:
I will go ahead with using stdout for this.

MylesBorins commented on lib/citgm-node-gyp.js:
```diff
@@ -0,0 +1,23 @@
+  fs.writeFileSync(marker_filename, 'node-gyp called');
+}
+
+delete process.env['npm_config_node_gyp'];
```
would prefer setting to null rather than delete. Is there a particular reason to use delete?

BridgeAR replied:
As not even though this is quite old: setting an environment variable to undefined will actually result in it being set to the string `'undefined'`.

MylesBorins commented on lib/create-options.js:
```diff
@@ -16,6 +18,20 @@ function createOptions(cwd, context) {
     });
   }
+
+  if (context.module && context.module.verifyNodeGyp !== undefined) {
```
I believe this can simply be `&& context.module.verifyNodeGyp`.

bzoz replied:
`verifyNodeGyp` can be `false`

MylesBorins commented on lib/create-options.js:
```diff
@@ -16,6 +18,20 @@ function createOptions(cwd, context) {
+    options.env['citgm_node_gyp_marker_filename'] = path.join(context.path,
+                                                              context.module.name,
+                                                              'citgm_marker');
+    if (process.env['npm_config_node_gyp']) {
```
we should likely document these environment variables in the readme

bzoz replied:
Ok

MylesBorins commented on lib/lookup.js:
```diff
@@ -103,6 +103,16 @@ function resolve(context, next) {
         context.module.script = rep.script;
       }
       context.module.flaky = context.options.failFlaky ? false : isMatch(rep.flaky);
+      if (rep['verify-node-gyp'] !== undefined) {
```
no need for `!== undefined` can simply be `if (rep['verify-node-gyp'])`

MylesBorins commented on lib/create-options.js:
```diff
@@ -16,6 +18,20 @@ function createOptions(cwd, context) {
   }
+
+  if (context.module && context.module.verifyNodeGyp !== undefined) {
+    options.env['citgm_node_gyp_marker_filename'] = path.join(context.path,
```
can this not be done with printing to stdout rather than putting it in a file?

bzoz replied:
It can. I'll change it.

MylesBorins commented on lib/npm/install.js:
```diff
@@ -82,6 +83,17 @@ function install(context, next) {
       return next(Error('Install Failed'));
     }
+    if (context.module.verifyNodeGyp !== undefined) {
+      var nodeGypCalled = fs.existsSync(path.join(context.path,
```
would really like this to not be done with Sync, in fact I feel like there shoudl be a way to do this through printing to stdout rather than relying on a file to pass data

bzoz commented on lib/citgm-node-gyp.js:
```diff
@@ -0,0 +1,23 @@
+  fs.writeFileSync(marker_filename, 'node-gyp called');
+}
+
+delete process.env['npm_config_node_gyp'];
```
Documentation (https://nodejs.org/api/process.html#process_process_env) says "use delete" so I went with that.

bzoz commented:
Rebased and updated. Now uses stdout to log if it was called. PTAL

@MylesBorins I think using env var to pass data to node-gyp is the best solution here. We are using those already to pass alternative node-gyp location to npm.

bzoz commented:
Rebased on master, PTAL

bzoz commented:
Rebased. @nodejs/citgm

BridgeAR commented:
This is LGTM. It just needs a rebase as far as I can tell.

MylesBorins commented on package.json:
```diff
@@ -36,6 +36,7 @@
     "columnify": "^1.5.1",
     "lodash": "^4.12.0",
     "mkdirp": "^0.5.1",
+    "node-gyp": "~3.4.0",
```
so it is possible that we are going to be testing on a version of node that has an older version of npm, and from that an older version of node-gyp... is this going to cause weird inconsistencies?

bzoz commented:
I think the original idea was to that https://github.com/nodejs/citgm/pull/319 would add testing compiling addons in CITGM.  Some of the packages do not accept `--build-from-source`  - I think `node-sass` was one of them, requiring setting up an env variable. So, this was for making sure that nothing in the package changed and `node-gyp` was called.

Since #319 was closed anyway, I don't think this is still needed.

bzoz commented:
I don't think it makes sense to keep this open. Closing.
